pkg/internal/source/kind.go

```diff
@@ -104,14 +177,69 @@ func (ks *Kind) String() string {
 // WaitForSync implements SyncingSource to allow controllers to wait with starting
 // workers until the cache is synced.
 func (ks *Kind) WaitForSync(ctx context.Context) error {
+	if func() bool {
+		ks.mu.Lock()
+		defer ks.mu.Unlock()
+
+		return ks.started == nil
+	}() {
+		return fmt.Errorf("cannot wait for sync on an unstarted source")
+	}
+
 	select {
 	case err := <-ks.started:
-		return err
+		if err != nil {
+			// Try to stop the source to clean up any resources
+			// and collect any errors.
+			if stopErr := ks.Stop(); stopErr != nil {
+				err = kerrors.NewAggregate([]error{err, fmt.Errorf("failed to stop source: %w", stopErr)})
+			}
+
+			return err
+		}
+
+		return nil
 	case <-ctx.Done():
-		ks.startCancel()
 		if errors.Is(ctx.Err(), context.Canceled) {
 			return nil
 		}
 		return errors.New("timed out waiting for cache to be synced")
 	}
 }
+
+// Stop is internal and should be called only by the Controller to stop the Source. It should block until the
+// Source has stopped.
+// Stop can be called multiple times, and should be idempotent. If the Source is not running, Stop will prevent
+// the Source from starting in the future without error.
```

Comment:
```suggestion
// Stop can be called multiple times, and should be idempotent. If the Source is not running, Stop will prevent
// the Source from starting in the future.
```

nit: otherwise it's misleading as it will just prevent the source from starting but not from "starting without error"